src/item.cpp

```diff
@@ -2354,76 +2271,252 @@ std::string item::tname( unsigned int quantity, bool with_prefix ) const
-    ret << string_format( _( "%1$s%2$s%3$s%4$s%5$s%6$s" ), damtext.c_str(), burntext.c_str(),
-                          modtext.c_str(), vehtext.c_str(), maintext.c_str(), tagtext.c_str() );
+    std::string final = string_format( _( "%1$s%2$s%3$s%4$s%5$s%6$s" ), damtext,
+                                    burntext, modtext, vehtext, maintext, tagtext );
 
     if( item_vars.find( "item_note" ) != item_vars.end() ) {
         //~ %s is an item name. This style is used to denote items with notes.
-        return string_format( _( "*%s*" ), ret.str().c_str() );
+        return string_format( _( "*%s*" ), final );
+    }
+
+    return final;
+}
+
+enum naming_style : int {
+    NAMING_STANDARD,            // Wrench
+    NAMING_WITH_CHARGES,        // 00 Shot (10), Flashlight (100)
+    NAMING_WITH_AMMO_TYPE,      // Glock 18 (15) (9x19mm)
+    NAMING_CONTAINER_LEADING,   // plastic bottle with clean water (2)
+    NAMING_CONTAINER_TRAILING   // clean water (2) in plastic bottle
+};
+
+naming_style item::get_naming_style() const
+{
+    if ( is_container() && !contents.empty() ) {
+        // Non-empty container-type items
+        return NAMING_CONTAINER_TRAILING;
+    } else if ( is_gun() || is_magazine() ) {
+        // Guns and magazines
+        return NAMING_WITH_AMMO_TYPE;
+    } else if ( is_armor() && !contents.empty() ){
+        // Armor and clothing with non-empty contents
+        return NAMING_CONTAINER_LEADING;
+    } else if ( is_bandolier() && !contents.empty() ) {
+        // Non-empty bandoliers and quivers
+        return NAMING_CONTAINER_TRAILING;
+    } else if ( !contents.empty() && contents.front().is_magazine() ) {
+        // Other items containing magazines
+        return NAMING_CONTAINER_TRAILING;
+    } else if ( ammo_capacity() > 0 ) {
+        // Other items with ammo capacity (like flashlights)
+        return NAMING_WITH_CHARGES;
+    } else if ( !contents.empty() ) {
+        // Any other item with non-empty contents
+        return NAMING_CONTAINER_TRAILING;
+    } else if ( is_book() && get_chapters() > 0 ) {
+        // Books with remaining chapters
+        return NAMING_WITH_CHARGES;
+    } else if ( count_by_charges() && !has_infinite_charges() ) {
+        // Items with non-ammo charges (like lighters)
+        return NAMING_WITH_CHARGES;
     } else {
-        return ret.str();
+        // Items that don't need their contents, charges, or ammo type
+        // displayed (everything else in the game)
+        return NAMING_STANDARD;
     }
 }
 
 std::string item::display_name( unsigned int quantity ) const
 {
-    std::string name = tname( quantity );
-    std::string sidetxt;
-    std::string amt;
+    // Anonymous enum used for readability
```

Comment:
What "Anonymous enum"? If you add lots of comments, you also need to update them often.

Reply:
I've left the [WIP] in the title of the pull request for exactly this sort of thing. I was working on the actual functionality of the code until late last night. I promise I won't remove the [WIP] until I've gone over the code thoroughly at the end.